File: Zero-1 KnapSack/knapsack_bnb.py
class Item:
    def __init__(self, weight, value, idx):
        self.weight = weight
        self.value = value
        self.idx = idx
        self.ratio = value / weight

class Node:
    def __init__(self, level, value, weight, bound, items):
        self.level = level  # Level in decision tree
        self.value = value  # Current value
        self.weight = weight  # Current weight
        self.bound = bound  # Upper bound
        self.items = items  # Selected items

class Knapsack:
    def __init__(self) -> None:
        pass

    def calculate_bound(self, node, N, capacity, items):
        """Calculate upper bound for a node using fractional knapsack"""
        if node.weight >= capacity:
            return 0

        value_bound = node.value
        weight = node.weight
        j = node.level + 1

        # Add items until capacity is reached
        while j < N and weight + items[j].weight <= capacity:
            weight += items[j].weight
            value_bound += items[j].value
            j += 1

        # Add fraction of next item if possible
        if j < N:
            value_bound += (capacity - weight) * items[j].ratio

        return value_bound

    def branch_and_bound(self, max_capacity, weights, values):
        """
        Solve 0/1 Knapsack using Branch and Bound
        Args:
            max_capacity: Maximum capacity of knapsack
            weights: Array of item weights
            values: Array of item values
        Returns:
            max_value: Maximum value possible
            selected_items: List of indices of selected items
        """
        N = len(weights)

        # Create items and sort by value/weight ratio
        items = [Item(weights[i], values[i], i) for i in range(N)]
        items.sort(key=lambda x: x.ratio, reverse=True)

        # Initialize queue with root node
        queue = []
        root = Node(-1, 0, 0, 0, [])
        root.bound = self.calculate_bound(root, N, max_capacity, items)
        queue.append(root)

        max_value = 0
        best_items = []
        nodes_explored = 0

        while queue:
            # Get node with highest bound
            curr = max(queue, key=lambda x: x.bound)
            queue.remove(curr)
            nodes_explored += 1

            # Skip if bound is less than current best
            if curr.bound <= max_value:
                continue

            level = curr.level + 1

            # Skip if we've processed all items
            if level >= N:
                continue

            # Include current item
            new_weight = curr.weight + items[level].weight
            if new_weight <= max_capacity:
                new_items = curr.items + [items[level].idx]
                include = Node(level, 
                             curr.value + items[level].value,
                             new_weight,
                             0,
                             new_items)
                include.bound = self.calculate_bound(include, N, max_capacity, items)

                if include.value > max_value and include.weight <= max_capacity:
                    max_value = include.value
                    best_items = include.items

                if include.bound > max_value:
                    queue.append(include)

            # Exclude current item
            exclude = Node(level,
                          curr.value,
                          curr.weight,
                          0,
                          curr.items)
            exclude.bound = self.calculate_bound(exclude, N, max_capacity, items)

            if exclude.bound > max_value:
                queue.append(exclude)

        best_items.sort()
        print("The max profit by branch and bound is " + str(max_value))
        print("The item selected are: " + str(best_items))
        

weights = [3, 4, 6, 5]
values = [2, 3, 1, 4]
k = Knapsack()
k.branch_and_bound(8, weights, values)

File: Zero-1 KnapSack/knapsack_bnb.cpp
#include<bits/stdc++.h>
using namespace std;

struct Item {
    int weight;
    int value;
    int idx;
    double ratio;  // value per unit weight
    
    Item(int w, int v, int i) : weight(w), value(v), idx(i) {
        ratio = static_cast<double>(v) / w;
    }
};

// Structure for nodes in Branch and Bound tree
struct Node {
    int level;          // Level of node in decision tree
    int value;          // curr value of knapsack
    int weight;         // curr weight of knapsack
    double bound;       // Upper bound of maximum possible value
    vector<int> items;  // Items included in curr node(indices)
    
    Node(int l, int v, int w) : level(l), value(v), weight(w), bound(0.0) {}
};


class Knapsack {
    double calculateBound(Node *node, int capacity, vector<Item> &items) {
        // If we've exceeded the capacity, this node is not valid
        if (node -> weight >= capacity) return 0;

        // Initialize bound with curr value
        double valueBound = node -> value;
        int currWeight = node -> weight;
        int j = node -> level + 1;

        // Add items until we exceed capacity
        while (j < items.size() && currWeight + items[j].weight <= capacity) {
            currWeight += items[j].weight;
            valueBound += items[j].value;
            j++;
        }

        // Add the fraction of the next item if possible
        if (j < items.size()) {
            valueBound += (capacity - currWeight) * items[j].ratio;
        }

        return valueBound;
    }

public:
    void branchAndBound(int maxCapacity, vector<int> &weights, vector<int> &values) {
        int n = weights.size();    
        // Create and sort items by value/weight ratio
        vector<Item> items;
        for (int i = 0; i < n; i++) {
            items.push_back(Item(weights[i], values[i], i));
        }
        sort(items.begin(), items.end(), [](const Item &a, const Item &b) { return a.ratio > b.ratio; });

        // Create a priority queue that will sort nodes by their bounds
        auto compare = [](Node *a, Node *b) { return a -> bound < b -> bound; };
        priority_queue<Node *, vector<Node *>, decltype(compare)> pq(compare);

        // Create root node
        Node *root = new Node(-1, 0, 0);
        root -> bound = calculateBound(root, maxCapacity, items);
        pq.push(root);

        // Initialize best solution
        int maxValue = 0;
        vector<int> bestItems;
        int nodesCount = 0;

        // Branch and Bound
        while(!pq.empty()) {
            // Get node with highest bound
            Node *curr = pq.top();
            pq.pop();
            nodesCount++;

            // Skip if bound is less than curr best solution
            if (curr -> bound <= maxValue) {
                delete curr;
                continue;
            }

            // Get next level
            int level = curr -> level + 1;

            // If we've processed all items
            if (level >= n) {
                delete curr;
                continue;
            }

            // Try including curr item
            int newWeight = curr -> weight + items[level].weight;
            if (newWeight <= maxCapacity) {
                Node *includeNode = new Node(level, 
                                            curr -> value + items[level].value,
                                            newWeight);
                includeNode -> items = curr->items;
                includeNode -> items.push_back(items[level].idx);
                includeNode -> bound = calculateBound(includeNode, maxCapacity, items);

                // Update best solution if needed
                if (includeNode -> value > maxValue) {
                    maxValue = includeNode -> value;
                    bestItems = includeNode -> items;
                }

                if (includeNode -> bound > maxValue) {
                    pq.push(includeNode);
                } else {
                    delete includeNode;
                }
            }

            // Try excluding curr item
            Node *excludeNode = new Node(level, curr -> value, curr -> weight);
            excludeNode -> items = curr -> items;
            excludeNode -> bound = calculateBound(excludeNode, maxCapacity, items);

            if (excludeNode -> bound > maxValue) {
                pq.push(excludeNode);
            } else {
                delete excludeNode;
            }

            delete curr;
        }

        // Sort the items before returning
        sort(bestItems.begin(), bestItems.end());
        cout<<"The max profit by branch and bound is "<<maxValue<<endl;
        cout<<"The items selected are: ";
        for(int i : bestItems) cout<<i<<" ";
        cout<<endl;
    }
};

int main() {
    vector<int> weights = {3, 4, 6, 5};
    vector<int> values = {2, 3, 1, 4};

    Knapsack k;
    k.branchAndBound(8, weights, values);
    
    return 0;
}